File: c++/Y.cc
#include "Y.h"
#include "objetoA.h"
#include "objetoC.h"
/*B* Y::foo(B* test){
    test->fooB();
    objetoB* prueba = new objetoB();
    return prueba;
}*/
/*C* Y::foo(B* test){    
    objetoC* prueba = new objetoC();
    prueba->fooC();
    return prueba;
}*/
// Si permite covarianza de al retornar C*
/*A* Y::foo(B* test){    
    objetoA* prueba = new objetoA();
    prueba->fooA();
    return prueba;
}*/
//No permite contravarianza al retornar A*
B* Y::foo(A* test){    
    objetoB* prueba = new objetoB();
    test->fooA();
    return prueba;
}
//No permite contravarianza al ingresar el valor A*
B* Y::foo(C* test){    
    objetoB* prueba = new objetoB();
    test->fooC();
    return prueba;
}

File: java/Y.java
public class Y extends X{
	
/*    public B foo(B test){
		test.fooB();
        return new objetoB();
    }
}*/
// Se puede ver que esta funcionando correctamente
	
	public A foo(B test){
		test.fooB();
		return new objetoA();
		}
    }
/* No lo permite error:
./Y.java:10: error: foo(B) in Y cannot override foo(B) in X
public A foo(B test){
         ^
return type A is not compatible with B
1 error
*/
